pkg/executor/join/semi_join_probe.go

```diff
@@ -0,0 +1,439 @@
+				// selected:  | T | F | T | F | T | T | F | F | F | T | F | F | T |
+				//              ▲                                   ▲
+				//              └ output index 0 here               |
+				//                                                  |
+				//                       index 0 was true before, we should not output it again
+				encounterTrue = true
+				j.selected[i] = false
+			} else {
+				encounterTrue = j.selected[i]
+				if encounterTrue {
+					j.skipRowIdxSet[currentGroupIdx] = struct{}{}
+				}
+			}
+
+			scannedGroupIdx = currentGroupIdx
+			continue
+		}
+
+		if encounterTrue {
+			// When we have encountered true in a group,
+			// we need to set the rest of selected values to false.
+			j.selected[i] = false
+			continue
+		}
+
+		encounterTrue = j.selected[i]
+		if encounterTrue {
+			j.skipRowIdxSet[currentGroupIdx] = struct{}{}
+		}
+	}
+}
+
+func (j *semiJoinProbe) removeMatchedProbeRow() {
+	for idx, selected := range j.selected {
+		if selected {
+			delete(j.processedProbeRowIdxSet, j.groupMark[idx])
+		}
+	}
+}
+
+func (j *semiJoinProbe) matchMultiBuildRows(joinedChk *chunk.Chunk, joinedChkRemainCap *int, isRightSideBuild bool) {
+	tagHelper := j.ctx.hashTableContext.tagHelper
+	meta := j.ctx.hashTableMeta
+	for j.matchedRowsHeaders[j.currentProbeRow] != 0 && *joinedChkRemainCap > 0 && j.matchedRowsForCurrentProbeRow < maxMatchedRowNum {
+		candidateRow := tagHelper.toUnsafePointer(j.matchedRowsHeaders[j.currentProbeRow])
+		if isKeyMatched(meta.keyMode, j.serializedKeys[j.currentProbeRow], candidateRow, meta) {
+			if isRightSideBuild || !meta.isCurrentRowUsedWithAtomic(candidateRow) {
+				j.appendBuildRowToCachedBuildRowsV1(j.currentProbeRow, candidateRow, joinedChk, 0, true)
+				j.matchedRowsForCurrentProbeRow++
+				*joinedChkRemainCap--
+
+				if isRightSideBuild {
+					j.groupMark = append(j.groupMark, j.currentProbeRow)
+				}
+			}
+		} else {
+			j.probeCollision++
+		}
+		j.matchedRowsHeaders[j.currentProbeRow] = getNextRowAddress(candidateRow, tagHelper, j.matchedRowsHashValue[j.currentProbeRow])
+	}
+
+	j.finishLookupCurrentProbeRow()
+}
+
+func (j *semiJoinProbe) concatenateProbeAndBuildRows(joinedChk *chunk.Chunk, sqlKiller *sqlkiller.SQLKiller, isRightSideBuild bool) error {
+	j.groupMark = j.groupMark[:0]
+	joinedChkRemainCap := joinedChk.Capacity()
+
+	for joinedChkRemainCap > 0 && (len(j.processedProbeRowIdxSet) > 0 || j.nextProcessProbeRowIdx < j.chunkRows) {
+		for probeRowIdx := range j.processedProbeRowIdxSet {
+			j.currentProbeRow = probeRowIdx
+			j.matchMultiBuildRows(joinedChk, &joinedChkRemainCap, isRightSideBuild)
+
+			if j.matchedRowsHeaders[probeRowIdx] == 0 {
+				delete(j.processedProbeRowIdxSet, probeRowIdx)
+			}
+
+			if joinedChkRemainCap == 0 {
+				break
+			}
+		}
+
+		for joinedChkRemainCap > 0 && j.nextProcessProbeRowIdx < j.chunkRows {
+			j.currentProbeRow = j.nextProcessProbeRowIdx
+			j.matchMultiBuildRows(joinedChk, &joinedChkRemainCap, true)
```

Comment:
> why the third argument is hardcoded as true?

It's a mistake, I will fix it. UT does not show any wrong because this mistake not affects correctness of result.